### PR TITLE
[Clang] [Parser] Improve diagnostic for `friend concept`

### LINKED ISSUE #45182: "friends can only be classes or functions" should also be diagnosed when trying to befriend a concept
|  |  |
| --- | --- |
| Bugzilla Link | [45837](https://llvm.org/bz45837) |
| Version | trunk |
| OS | All |
| CC | @mizvekov,@zygoloid |

## Extended Description 
https://godbolt.org/z/jntYnj

Given this code:

struct S {
    template<class>
    friend concept fooable;
};

Clang produces this harsh syntax error:

error: expected member name or ';' after declaration specifiers
    friend concept fooable;
           ^

If you try to befriend any other kind of entity -- for example, a variable or variable template -- Clang gives a much friendlier error message, indicating that it understands what you were *trying* to do:

error: friends can only be classes or functions
    friend bool fooable_v;
                ^

People are already trying to befriend concepts in the field, so it would be nice if Clang gave that same friendly error message (diag::err_unexpected_friend) for the `concept` case, as well.

### PR DESCRIPTION
**Pull request by Sirraide**
Diagnose this early after parsing declaration specifiers; this allows us to issue a better diagnostic. This also checks for `concept friend` and concept declarations w/o a template-head because it’s easiest to do that at the same time.

Fixes #45182.

### CODE REVIEW

**Comment by Sirraide** [comment]
`template<class> concept friend fooable;` still gives the error ‘expected unqualified-id’; I could move the check for whether it’s a concept declaration at class scope to the start of `ParseConceptDefinition` instead of doing that in `ActOnStartConceptDefinition` or whatever it was called and add a check for whether it’s friend-specified, but I’m not sure that’s really necessary.

**Review by cor3ntin — APPROVED** [review]
LGTM

File: clang/lib/Parse/ParseDeclCXX.cpp [hunk 1]
```diff
@@ -3139,6 +3139,19 @@ Parser::DeclGroupPtrTy Parser::ParseCXXClassMemberDeclaration(
     return Actions.BuildDeclaratorGroup(Decls);
   }
 
+  // Befriending a concept is invalid and would already fail if
+  // we did nothing here, but this allows us to issue a more
+  // helpful diagnostic.
+  if (Tok.is(tok::kw_concept)) {
```
Comment (shafik) [inline clang/lib/Parse/ParseDeclCXX.cpp:3145]: Could we get a standard reference for a befriending a concept being invalid.

Comment (Sirraide) [reply clang/lib/Parse/ParseDeclCXX.cpp:3145]: > Could we get a standard reference for a befriending a concept being invalid.

Pretty sure the syntax just doesn’t allow for it: Neither a `simple-type-specifier`, nor an `elaborated-type-specifier`, nor a `typename-specifier` can start with `concept`. The grammatical productions for `friend-type-specifier` are already listed in the comment at the top of the function.

**Comment by Sirraide** [comment]
> LLVM Buildbot has detected a new failure on builder `openmp-offload-libc-amdgpu-runtime`

Er, if that’s not unrelated then I have no idea what’s going on. I’m pretty sure we can ignore that.

**Comment by cor3ntin** [comment]
yup. Post commit CI is batched, someone's else PR broke it

**Comment by Sirraide** [comment]
> yup. Post commit CI is batched, someone's else PR broke it

Yeah, that’s what I figured, but normally the bot doesn’t comment on the pr if that’s the case—or at least it didn’t use to do that.

**Pull request merged**